components/weekly_storage/weekly_event_storage_unittest.cc

```diff
@@ -0,0 +1,118 @@
+    kBrave,
+    kMaxValue = kBrave,
+  };
+
+  WeeklyEventStorageTest() : clock_(new base::SimpleTestClock) {
+    pref_service_.registry()->RegisterListPref(kPrefName);
+
+    state_ = std::make_unique<WeeklyEventStorage<TestValues>>(
+        &pref_service_, kPrefName, std::unique_ptr<base::Clock>(clock_));
+    clock_->SetNow(base::Time::Now());
+  }
+
+ protected:
+  base::SimpleTestClock* clock_;
+  TestingPrefServiceSimple pref_service_;
+  std::unique_ptr<WeeklyEventStorage<TestValues>> state_;
+};
+
+TEST_F(WeeklyEventStorageTest, StartsEmpty) {
+  EXPECT_FALSE(state_->HasEvent());
+  EXPECT_EQ(state_->GetLatest(), absl::nullopt);
+}
+
+TEST_F(WeeklyEventStorageTest, AddEvents) {
+  state_->Add(TestValues::kNull);
+  EXPECT_TRUE(state_->HasEvent());
+  EXPECT_EQ(state_->GetLatest(), absl::optional<TestValues>(TestValues::kNull));
+
+  state_->Add(TestValues::kBrave);
+  EXPECT_EQ(state_->GetLatest(),
+            absl::optional<TestValues>(TestValues::kBrave));
+}
+
+TEST_F(WeeklyEventStorageTest, ForgetsOldEvents) {
+  // Add an initial event.
+  state_->Add(TestValues::kFoo);
+  EXPECT_EQ(state_->GetLatest(), absl::optional<TestValues>(TestValues::kFoo));
+
+  // Jump to the next week.
+  clock_->Advance(base::TimeDelta::FromDays(8));
+  // Should have forgotten about older days.
+  EXPECT_FALSE(state_->HasEvent());
+
+  // Newer events should still accumulate.
+  state_->Add(TestValues::kNull);
+  state_->Add(TestValues::kBar);
+  EXPECT_EQ(state_->GetLatest(), absl::optional<TestValues>(TestValues::kBar));
+}
+
+TEST_F(WeeklyEventStorageTest, IntermittentUsage) {
+  auto value = TestValues::kFoo;
+  for (int day = 0; day < 10; day++) {
+    clock_->Advance(base::TimeDelta::FromDays(day % 3));
+    state_->Add(value);
+  }
+  EXPECT_EQ(state_->GetLatest(), absl::optional<TestValues>(value));
+}
+
+TEST_F(WeeklyEventStorageTest, InfrequentUsage) {
+  state_->Add(TestValues::kFoo);
+  clock_->Advance(base::TimeDelta::FromDays(6));
+  state_->Add(TestValues::kBar);
+  EXPECT_EQ(state_->GetLatest(), absl::optional<TestValues>(TestValues::kBar));
+  clock_->Advance(base::TimeDelta::FromDays(10));
+  EXPECT_EQ(state_->GetLatest(), absl::nullopt);
+}
+
+// Verify serialization order across reloads, since GetLatest
+// relies on this.
+// FIXME(rillian): Disabled because it crashes as written.
+TEST_F(WeeklyEventStorageTest, DISABLED_SerializationOrder) {
+  // Add a series of events.
+  state_->Add(TestValues::kFoo);
+  state_->Add(TestValues::kBar);
+  clock_->Advance(base::TimeDelta::FromDays(1));
+  state_->Add(TestValues::kFoo);
+  state_->Add(TestValues::kBrave);
+  clock_->Advance(base::TimeDelta::FromDays(1));
+  EXPECT_EQ(state_->GetLatest(),
+            absl::optional<TestValues>(TestValues::kBrave));
+
+  // Drop the WeeklyEventStorage object.
+  state_.reset(nullptr);
+
+  // Create a new one.
+  state_ = std::make_unique<WeeklyEventStorage<TestValues>>(
+      &pref_service_, kPrefName, std::unique_ptr<base::Clock>(clock_));
```

Comment:
I think you need to create a new `clock_` instance here, because the old one is destroyed.